collector/filesystem_freebsd.go

```diff
@@ -0,0 +1,88 @@
+// Copyright 2015 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// +build !nofilesystem
+
+package collector
+
+import (
+	"bytes"
+	"unsafe"
+
+	"github.com/prometheus/common/log"
+	"golang.org/x/sys/unix"
+)
+
+const (
+	defIgnoredMountPoints = "^/(dev)($|/)"
+	defIgnoredFSTypes     = "^devfs$"
+	MNT_RDONLY            = 0x1
+	MNT_NOWAIT            = 0x2
+)
+
+func gostring(b []int8) string {
+	bb := *(*[]byte)(unsafe.Pointer(&b))
+	idx := bytes.IndexByte(bb, 0)
+	if idx < 0 {
+		return ""
+	}
+	return string(bb[:idx])
+}
+
+// Expose filesystem fullness.
+func (c *filesystemCollector) GetStats() (stats []filesystemStats, err error) {
+	buf := make([]unix.Statfs_t, 16)
+	for {
+		n, err := unix.Getfsstat(buf, MNT_NOWAIT)
+		if err != nil {
+			return nil, err
+		}
+		if n < len(buf) {
+			buf = buf[:n]
+			break
+		}
+		buf = make([]unix.Statfs_t, len(buf)*2)
+	}
+	stats = []filesystemStats{}
+	for _, fs := range buf {
+		mountpoint := gostring(fs.Mntonname[:])
+		if c.ignoredMountPointsPattern.MatchString(mountpoint) {
+			log.Debugf("Ignoring mount point: %s", mountpoint)
+			continue
+		}
+
+		device := gostring(fs.Mntfromname[:])
+		fstype := gostring(fs.Fstypename[:])
+		if c.ignoredFSTypesPattern.MatchString(fstype) {
+			log.Debugf("Ignoring fs type: %s", fstype)
+			continue
+		}
+
+		var ro float64
+		if (fs.Flags & MNT_RDONLY) != 0 {
```

Comment:
The suggested alternative looks like a blob of ink to me, precisely because of how gofmt formats it. Parentheses also help visualize the order of operations, but render much more readable in the case of bitmasks.